### PR TITLE
Simplify build infrastructure

### LINKED ISSUE #611: Drop support for Node.js 6.x
Considering that Node.js 6.x will enter maintenance mode long time before our expected GA release, we think it makes sense to require LoopBack Next users to run on Node.js 8.x or newer.

**Tasks**

 - [x] Drop Node.js 6.x from Travis-CI and AppVeyor configs
 - [x] Change "engines" field in `package.json` to `"node": ">=8"` 
 - [x] Simplify build-related tasks in both top-level and package-level `package.json` files - drop `dist6` - related targets.
 - [x] Replace all `promisify` helpers with `util.promisify` from Node.js core
 - [x] Simplify `@loopback/build`, remove scripts dealing with multiple `dist` folders and multiple `tsc` targets (es2015, es2017).
 - [x] Consider removing top-level `index*` files in each package, set `"main": "dist/index.js"`  in `package.json` instead. (This may not work.)
 - [x] Update the project template used by CLI when scaffolding new projects
 - [x] Apply the same changes to all other related repositories + review READMEs and remove mentions of Node.js 6.x:

   - [x] https://github.com/strongloop/loopback-next-hello-world
   - [x] https://github.com/strongloop/loopback4-extension-starter
   - [x] https://github.com/strongloop/loopback4-example-microservices
   - [x] https://github.com/strongloop/loopback4-example-getting-started
   - [x] https://github.com/strongloop/loopback4-extension-grpc
   - [x] https://github.com/strongloop/loopback4-extension-typeorm
   - [x] https://github.com/strongloop/loopback4-extension-mqtt
   - [x] https://github.com/strongloop/loopback4-extension-graphql
   - [x] https://github.com/strongloop/loopback4-example-log-extension
   - [x] https://github.com/strongloop/loopback4-example-rpc-server
   - [x] https://github.com/strongloop/loopback4-example-hello-world

 - [x] Review documentation on loopback.io and update all places referring to Node.js 6.x
  
 - [x] Write a blog post

### PR DESCRIPTION
**Pull request by b-admike**
Related to #611 

## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] Related API Documentation was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `packages/example-*` were updated

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "build": "lerna run build --parallel --loglevel=silent",
     "build:current": "lerna run build:current --parallel --loglevel=silent",
```
Comment (shimks) [inline package.json]: Is there still need for this?

File: packages/cli/generators/project/templates/package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "loopback-<%= project.projectType -%>",
     "loopback"
   ],
-  "main": "index.js",
+  "main": "dist/index.js",
   "engines": {
     "node": ">=8"
   },
```
Comment (shimks) [inline packages/cli/generators/project/templates/package.json]: remove `dist6` from `files` (way down; github won't let me comment on unchanged lines)

File: packages/core/package.json [hunk 3]
```diff
@@ -38,7 +35,6 @@
     "index.js",
     "index.d.ts",
```
Comment (shimks) [inline packages/core/package.json]: Do we still need this (and `index.js`)?

File: package.json (hunk 1 above)
Comment (b-admike) [reply package.json]: nope, good catch!

File: packages/core/package.json (hunk 3 above)
Comment (b-admike) [reply packages/core/package.json]: nope :).

File: packages/build/bin/utils.js [hunk 4]
```diff
@@ -12,11 +12,12 @@ const spawn = require('cross-spawn');
 const debug = require('debug')('loopback:build');
 
 /**
- * Get the Node.js compilation target - es2017 or es2015
+ * Get the Node.js compilation target - es2017
```
Comment (raymondfeng) [inline packages/build/bin/utils.js]: IMO, we should keep `es2015` support for `@loopback/build` as it can be used to build TypeScript projects independent of `LoopBack 4`.

File: packages/cli/generators/project/templates/package.json [hunk 5]
```diff
@@ -52,10 +52,8 @@
   "license": "MIT",
   "files": [
     "README.md",
-    "index.js",
-    "index.d.ts",
-    "dist",
-    "dist6"
+    "dist/index.js",
```
Comment (raymondfeng) [inline packages/cli/generators/project/templates/package.json]: Is it redundant with `dist`?

File: packages/core/package.json [hunk 6]
```diff
@@ -26,19 +24,16 @@
...
   },
   "devDependencies": {
     "@loopback/build": "^4.0.0-alpha.12",
     "@loopback/testlab": "^4.0.0-alpha.22"
   },
   "files": [
     "README.md",
-    "index.js",
-    "index.d.ts",
+    "dist/index.js",
```
Comment (raymondfeng) [inline packages/core/package.json]: Please note we have the following files under `dist`. We should include all of them except `test`. The top level `index.*` files are useful for source map.
```
index.d.ts
index.js
index.js.map
src
test
```

File: .travis.yml [hunk 7]
```diff
@@ -24,8 +27,8 @@ jobs:
       before_script: skip
       script: /bin/bash ./bin/lint-commits.sh
       after_success: skip
-      node_js:
-        - "8"
+      os:
```
Comment (jannyHou) [inline .travis.yml]: any reason why `osx` is not in the list?

Comment (virkt25) [reply .travis.yml]: Commit message linting only needs to happen once (on any platform) so for this stage we explicitly specify linux. Not specifying any will trigger commit message linting on both platforms which is redundant.

File: .travis.yml [hunk 8]
```diff
@@ -1,9 +1,12 @@
 sudo: false
 language: node_js
 node_js:
-  - "6"
   - "8"
 
+os:
+  - linux
+  - osx
```
Comment (bajtos) [inline .travis.yml]: Please don't make two changes in one pull request. Adding `osx` to our build matrix should be done in a standalone pull request.

File: packages/authentication/index.js [hunk 9]
```diff
@@ -1,7 +0,0 @@
-// Copyright IBM Corp. 2017,2018. All Rights Reserved.
-// Node module: @loopback/authentication
-// This file is licensed under the MIT License.
-// License text available at https://opensource.org/licenses/MIT
-
-const nodeMajorVersion = +process.versions.node.split('.')[0];
-module.exports = nodeMajorVersion >= 7 ? require('./dist') : require('./dist6');
```
Comment (bajtos) [inline packages/authentication/index.js]: With Node.js 10.0.0 coming in few months, I think we should start planning how to transpile to es2018 target. Since transpiling for both Node.js 8.x and 10.x will require these index files around, I think we should preserve them.

https://twitter.com/jasnell/status/958765477089632257
> PSA: Mark your calendars, the expected release date for Node.js 10.0.0 is on or around April 24th. Expect the first As-Close-To-A-Release-Candidate-As-We-Can-Get-It to hit in early April.

Thoughts?

File: packages/authentication/package.json [hunk 10]
```diff
@@ -3,19 +3,17 @@
   "version": "4.0.0-alpha.30",
   "description": "A LoopBack component for authentication support.",
   "engines": {
-    "node": ">=6"
+    "node": ">=8"
   },
+  "main": "dist/index.js",
```
Comment (bajtos) [inline packages/authentication/package.json]: ❗️ This is critical ❗️ 

I think this will break monorepo-wide refactorings in VSCode. Before your change, the TypeScript server was seeing `packages/authentication/index.ts` and ignoring any `.js`/`.d.ts` files produced by our build. With your change in place, the TypeScript server will see `packages/authentication/dist/index.js` (and `index.d.ts`), which will hide the original TypeScript sources.

Here are things that will stop working if my assessment is correct:

 - "Go to definition" across packages (e.g. find usage of `@inject` in `authentication` package, press F12) will jump to `.d.ts` file in `dist` instead of the original `.ts` file in `src`.

 - Refactorings like "rename" will not change all places using the renamed entity. Two different scenarios to verify: rename at the place where the entity is defined, rename at the place where the entity is used.

Please verify whether the three scenarios mentioned above work with your change in place. If they do, then my understanding of TypeScript integration in VSCode is no longer correct and you can ignore my comment.

File: packages/build/bin/utils.js (hunk 4 above)
Comment (bajtos) [reply packages/build/bin/utils.js]: +1

Also as I mentioned above, I'd like us to add es2018 target soon (around April this year, when Node.js 10.0.0 is released).

 - V8 6.4 already supports all ES2018 features according to [this tweet](https://twitter.com/v8js/status/956976954074267648)
 - Node's `master` branch contains V8 6.4 AFAICT, see [deps/v8](https://github.com/nodejs/node/tree/936eea5099bb5ab31b76ef22fcf15baed0cdf5ab/deps/v8)

File: packages/cli/generators/app/templates/index.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+// DO NOT EDIT THIS FILE
+// Add any aditional (re)exports to src/index.ts instead.
+export * from './src';
+
+const application = (module.exports = require('./src'));
```
Comment (bajtos) [inline packages/cli/generators/app/templates/index.ts]: My two comment above advocating for preservation of package-level `index.js` file apply to this place too.

File: packages/authentication/index.js (hunk 9 above)
Comment (bajtos) [reply packages/authentication/index.js]: (I do realize this contradicts the checklist in https://github.com/strongloop/loopback-next/pull/945, which I created back in October last year.)

File: packages/build/bin/utils.js (hunk 4 above)
Comment (bajtos) [reply packages/build/bin/utils.js]: (I do realize this contradicts the checklist in https://github.com/strongloop/loopback-next/pull/945, which I created back in October last year.)

File: packages/authentication/index.js (hunk 9 above)
Comment (bajtos) [reply packages/authentication/index.js]: After thinking about this a bit more: we should leave out transpilation target `es2018` out of scope of this pull request.

If we need to keep `index.js` file around because of https://github.com/strongloop/loopback-next/pull/945#discussion_r165288862, then I am ok with simplifying it to a simple redirect to `dist`. If it turns out TS+VSCode works well without this hand-written `index.js` around, then I am ok with deleting it as is proposed here.

**Comment by bajtos** [comment]
> Closes #611

@b-admike This phrase is telling GitHub to close #611 automatically after this pull request is landed. Are you sure all other tasks in #611 will be done by that time? In my experience, it's usually better to close user stories manually and avoid using automatic closing via "closes" or "fixes".

File: packages/authentication/package.json (hunk 10 above)
Comment (b-admike) [reply packages/authentication/package.json]: Good catch! The first scenario works, but the other two do not. Would ignoring those files in their new locations fix this issue? (will try and check back)

File: .travis.yml (hunk 8 above)
Comment (b-admike) [reply .travis.yml]: The historic reason for this addition is that Travis was only running commit linter instead of our build, since there was no longer a matrix with https://github.com/strongloop/loopback-next/pull/945/commits/1e5bc766ff6e85a4ebb20a39c0b6b35826099294. TV and I added it to address that issue. I agree that is a change on its own, so I can create a PR for it.

File: packages/authentication/index.js (hunk 9 above)
Comment (b-admike) [reply packages/authentication/index.js]: Actually yeah we were discussing about es2018 before making this change. I think it might be better to keep them as they do no harm, and if used in the future, we don't have to remake them.

**Comment by b-admike** [comment]
@bajtos Sorry, should not have used `closes` phrase. There are still outstanding checkboxes in that task that I need to make PRs for in their respective repos.

File: .travis.yml (hunk 8 above)
Comment (bajtos) [reply .travis.yml]: Wow, I did not realize a single Node.js target would break the matrix build. Thanks for explaining this to me. Your other PR looks good to me. Another option we can consider is to add Node.js 9 to the list of platforms we are testing.

File: packages/core/package.json (hunk 3 above)
Comment (bajtos) [reply packages/core/package.json]: See https://github.com/strongloop/loopback-next/pull/945#discussion_r165288862

File: packages/authentication/index.js [hunk 12]
```diff
@@ -3,5 +3,6 @@
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
+/* eslint-disable no-unused-vars*/
 const nodeMajorVersion = +process.versions.node.split('.')[0];
```
Comment (bajtos) [inline packages/authentication/index.js]: I think we should remove `nodeMajorVersion` if it's not used (instead of adding `eslit-disable` comment).

The same comment applies to all other places below where you added `eslint-disable`.

(BTW I believe we don't have `eslint` configured in our monorepo, do you have `eslint` configured in your IDE independently of our project setup?)

Comment (b-admike) [reply packages/authentication/index.js]: Okay, will do. Yeah I have `eslint`, but realized we use `tslint` after, and I have that  change in the other PRs :).
